From a7ac6185290ce704734f3c0cf107153c92fdaa34 Mon Sep 17 00:00:00 2001
From: Richard Liu <r66033@freescale.com>
Date: Thu, 11 Jul 2013 13:46:40 +0800
Subject: [PATCH] ENGR00270597 fix skia crash when rotate board in Launcher
 APPS

system will return to main menu after rotating SDP board several times
in Launcher APPS menu, you can see skia crash log in console, root cause
is there is some NULL function pointer, fix the crash throught bypass the
NULL function call.

Signed-off-by: Richard Liu <r66033@freescale.com>

Conflicts:
	src/core/SkBitmapProcState_sample.h
---
 src/core/SkBitmap.cpp               | 2 ++
 src/core/SkBitmapProcShader.cpp     | 5 +++++
 src/core/SkBitmapProcState_sample.h | 3 +++
 3 files changed, 10 insertions(+)

diff --git a/src/core/SkBitmap.cpp b/src/core/SkBitmap.cpp
index 6f90b71..cab249f 100644
--- a/src/core/SkBitmap.cpp
+++ b/src/core/SkBitmap.cpp
@@ -162,6 +162,8 @@ void SkBitmap::updatePixelsFromRef() const {
                 p = (char*)p
                     + fPixelRefOrigin.fY * fRowBytes
                     + fPixelRefOrigin.fX * fInfo.bytesPerPixel();
+            } else {
+                return;
             }
             fPixels = p;
             fColorTable = fPixelRef->colorTable();
diff --git a/src/core/SkBitmapProcShader.cpp b/src/core/SkBitmapProcShader.cpp
index 9d18119..113c0f0 100644
--- a/src/core/SkBitmapProcShader.cpp
+++ b/src/core/SkBitmapProcShader.cpp
@@ -193,6 +193,11 @@ void SkBitmapProcShader::BitmapProcShaderContext::shadeSpan(int x, int y, SkPMCo
     SkASSERT(state.fBitmap->getPixels());
     SkASSERT(state.fBitmap->pixelRef() == NULL ||
              state.fBitmap->pixelRef()->isLocked());
+    SkASSERT(state.fMatrixProc);
+    SkASSERT(state.fSampleProc32);
+
+    if (!mproc || !sproc || !state.fBitmap->getPixels())
+        return;
 
     for (;;) {
         int n = count;
diff --git a/src/core/SkBitmapProcState_sample.h b/src/core/SkBitmapProcState_sample.h
index 7030516..85d49f3 100644
--- a/src/core/SkBitmapProcState_sample.h
+++ b/src/core/SkBitmapProcState_sample.h
@@ -86,6 +86,7 @@ void MAKENAME(_nofilter_DX)(const SkBitmapProcState& s,
     SkASSERT(count > 0 && colors != NULL);
     SkASSERT(s.fInvType <= (SkMatrix::kTranslate_Mask | SkMatrix::kScale_Mask));
     SkASSERT(kNone_SkFilterQuality == s.fFilterLevel);
+    SkASSERT(s.fBitmap->getPixels());
     SkDEBUGCODE(CHECKSTATE(s);)
 
 #ifdef PREAMBLE
@@ -93,6 +94,8 @@ void MAKENAME(_nofilter_DX)(const SkBitmapProcState& s,
 #endif
     const SRCTYPE* SK_RESTRICT srcAddr = (const SRCTYPE*)s.fBitmap->getPixels();
 
+    if (srcAddr == NULL)
+        return;
     // buffer is y32, x16, x16, x16, x16, x16
     // bump srcAddr to the proper row, since we're told Y never changes
     SkASSERT((unsigned)xy[0] < (unsigned)s.fBitmap->height());
-- 
1.8.0

